### PR TITLE
webhooks: Initial impl

### PR DESCRIPTION
**Pull request by ParkMyCar**
### Motivation

Progress towards: https://github.com/MaterializeInc/materialize/issues/19951

Closes: #20210 

An initial implementation of webhook sources. There are still a number of things missing, but this PR does the following:

* Adds a new API to `StorageController` so we can append to a managed collection.
* Updates the SQL parser to support a new syntax.
* Support creating a webhook source in the Adapter.
* New HTTP endpoint that users can push data to, and a new Coordinator command that can be used to append rows.
* A single LaunchDarkly flag that gates creating a webhook source and appending data.
* Adds a new `webhook-append` command to testdrive to exercise the new source.

Notably missing from this PR is validation of requests, that will be in a followup.

### Tips for reviewer

I broke this diff up into logical sections to make reviewing easier:

1. Changes to `StorageController` for the new API
2. Updates to the SQL parser
3. Changes in the Adapter to support creating a new source
4. New HTTP endpoint in `environmentd`, and the new coordinator command
5. Testing. Parser tests, Rust tests, sqllogictests, new testdrive command and testdrive tests

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Adds the ability to create a webhook source, gated under a LaunchDarkly flag because it's not complete yet.

### CODE REVIEW

File: src/adapter/src/coord/command_handler.rs
Comment (def-) [inline src/adapter/src/coord/command_handler.rs:76]: Is this tested somewhere?

File: src/adapter/src/coord/command_handler.rs
Comment (def-) [inline src/adapter/src/coord/command_handler.rs:785]: Could test that blocking this actually works.

File: src/environmentd/src/http/webhook.rs
Comment (def-) [inline src/environmentd/src/http/webhook.rs:91]: Could test with another type to see that this error comes out?

File: src/environmentd/src/http/webhook.rs
Comment (def-) [inline src/environmentd/src/http/webhook.rs:153]: Can we test this error state?

File: src/environmentd/src/http/webhook.rs
Comment (def-) [inline src/environmentd/src/http/webhook.rs:142]: Can we test this error state?

**Review by def- — COMMENTED** [review]
Nightly looked fine: https://buildkite.com/materialize/nightlies/builds/2666#0188fecf-c5dd-4865-8a92-a0c768a48bf7
Coverage: https://buildkite.com/materialize/coverage/builds/161 Maybe some further error cases could be covered.
I'll write a platform check and add the commit to this PR.

**Review by def- — COMMENTED** [review]
Is this expected when trying to print the header column?
```
! SELECT * FROM webhook_json_with_headers;
contains: binary encoding of map types is not implemented
```

File: src/adapter/src/coord/command_handler.rs
Comment (ParkMyCar) [reply src/adapter/src/coord/command_handler.rs:76]: It is not, AFAICT this method is only called if we fail RBAC, and there aren't any RBAC restrictions on appending to a webhook, so this code path isn't used

File: src/environmentd/src/http/webhook.rs
Comment (ParkMyCar) [reply src/environmentd/src/http/webhook.rs:91]: Added some unit tests in this file to cover this scenario

File: src/adapter/src/coord/command_handler.rs
Comment (ParkMyCar) [reply src/adapter/src/coord/command_handler.rs:785]: Added a testdrive scenario that should cover this!

**Comment by ParkMyCar** [comment]
> Is this expected when trying to print the headers column?
> 
> ```
> ! SELECT * FROM webhook_json_with_headers;
> contains: binary encoding of map types is not implemented
> ```

AFAIK that is expected 😔 The `map` type we use for headers is MZ specific, and we did not implement the binary encoding for it, which testdrive uses.

Also, sorry I blew away the additional test coverage you added! I missed it when I last rebased off `main` and pushed, it should all be merged in now though. Thanks for the help!

File: src/sql-parser/src/parser.rs
Comment (sploiselle) [inline src/sql-parser/src/parser.rs:2468]: Prepositions are hard. I will think about this, as well, but `FROM` _feels_ wrong here to me. `AS` makes a little more sense but I'm not convinced. If you come up with an argument one way or another, lmk.

File: src/sql-parser/src/parser.rs
Comment (sploiselle) [inline src/sql-parser/src/parser.rs:2642]: I would suggest using `parse_format()` and then erroring in planning if you can an incompatible format--you're going to need to do that check anyway and I think it's more sane to do that with an error rather than an `unreachable`.

File: src/storage-client/src/controller.rs
Comment (sploiselle) [inline src/storage-client/src/controller.rs:189]: Once we have validation in webhooks, we'll need to determine if it has any non-system dependencies, e.g. someone used one of their own custom types in the body. Nothing to do now but something to be aware of.

File: src/adapter/src/catalog.rs
Comment (sploiselle) [inline src/adapter/src/catalog.rs:2081]: You know what the consequence of this state is? There's a chance this isn't "true" because we are not e.g. using a `CONNECTION`.

File: src/adapter/src/coord/sequencer/inner.rs
Comment (sploiselle) [inline src/adapter/src/coord/sequencer/inner.rs:195]: Can you add a TODO or a link here to the GitHub epic for webhooks? Eventually we will definitively want webhooks to take this path, e.g. once they're running in dataflows.

File: src/adapter/src/coord/sequencer/inner.rs
Comment (sploiselle) [inline src/adapter/src/coord/sequencer/inner.rs:244]: What is the advantage of running this in its own function instead of `sequence_create_source`?

File: src/sql/src/plan/statement/ddl.rs
Comment (sploiselle) [inline src/sql/src/plan/statement/ddl.rs:410]: Oh yeah, like I was saying in the parser reivew, you're already erroring here so you can be more permissive in the parser

File: src/sql/src/session/vars.rs
Comment (sploiselle) [inline src/sql/src/session/vars.rs:1118]: This sentence feels off to me (my undergrad is in English, sorry)--would you be open to something like:

> creating or pushing data to webhook sources

**Comment by sploiselle** [comment]
https://github.com/MaterializeInc/materialize/pull/20171#issuecomment-1611611587

Yes, expected. Binary decoding our non-PG types is a real rat's nest.

File: src/environmentd/src/http/webhook.rs
Comment (jkosh44) [inline src/environmentd/src/http/webhook.rs:169]: [405](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/405) or [415](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/415) might be a more appropriate error code. I think 403 Forbidden indicates that the request would be possible if the client had the right privileges.

**Review by jkosh44 — COMMENTED** [review]
Looking really good!

File: src/adapter/src/coord/sequencer/inner.rs
Comment (sploiselle) [reply src/adapter/src/coord/sequencer/inner.rs:306]: Think this should just be "source"; these are all something like `ObjectType`'s `to_string` implementation.

File: src/storage-client/src/controller/collection_mgmt.rs
Comment (ParkMyCar) [reply src/storage-client/src/controller/collection_mgmt.rs]: Makes sense! `mpsc::Sender` already implemented `Debug`, so I got rid of this custom impl and just used `#[derive(...)]` instead

File: src/sql-parser/src/parser.rs
Comment (ParkMyCar) [reply src/sql-parser/src/parser.rs:2468]: I agree, `FROM` does feel weird here, but the reason I think we should stick with `FROM` is to be consistent with other sources. IMO it would be even _more_ weird if `WEBHOOK` was the only source that used `AS` instead of `FROM`. But I'm totally open to using `AS` if that's the consensus

File: src/sql-parser/src/parser.rs
Comment (ParkMyCar) [reply src/sql-parser/src/parser.rs]: Originally I used `parse_format()` but removed it, because the errors it returned were misleading. For example, if a user ran a query like...

```
CREATE SOURCE foo FROM WEBHOOK BODY FORMAT AVRO
```

...the returned error was that it expected a keyword `USING`, and it would keep returning parsing errors until you provided an accurate AVRO schema. But in the end the user would get a planning error anyways. It's a small thing, but it seemed like a bad user experience of "stumbling" through the parser only to then also get rejected by planning.

Refactored to use `self.expect_one_of_keywords(...)`, which is 100% cleaner, thanks for the tip!

File: src/storage-client/src/controller.rs
Comment (ParkMyCar) [reply src/storage-client/src/controller.rs:189]: Good callout! I added a TODO here

File: src/adapter/src/coord/sequencer/inner.rs
Comment (ParkMyCar) [reply src/adapter/src/coord/sequencer/inner.rs:252]: The reason I split this out was `sequence_create_source` takes a `CreateSourcePlan` which requires a `SourceSinkClusterConfig`. Instead of trying to change one of these pieces to represent a `None` cluster, I figured it was easier to split the logic. I added a TODO here to merge/delete the method once we run webhook sources on `clusterd`

File: src/sql/src/plan/statement/ddl.rs
Comment (ParkMyCar) [reply src/sql/src/plan/statement/ddl.rs:410]: Commented above in more detail, but I wanted to be more restrictive in the parser for better error messages. Let me know if you want me to change this 🙂

File: src/sql/src/session/vars.rs
Comment (ParkMyCar) [reply src/sql/src/session/vars.rs]: No worries at all! English class was never my strongest 😁 It's definitely a bad sentence, I updated it to your suggestion!

File: src/sql-parser/src/ast/defs/statement.rs
Comment (ParkMyCar) [reply src/sql-parser/src/ast/defs/statement.rs]: Updated!

File: src/sql-parser/src/parser.rs
Comment (ParkMyCar) [reply src/sql-parser/src/parser.rs]: Yeah this was a stale comment from when I was prototyping, removed!

File: src/sql-parser/src/parser.rs
Comment (ParkMyCar) [reply src/sql-parser/src/parser.rs]: 🦅👀

File: src/sql-parser/tests/testdata/create
Comment (ParkMyCar) [reply src/sql-parser/tests/testdata/create:1]: Added!

File: src/adapter/src/rbac.rs
Comment (ParkMyCar) [reply src/adapter/src/rbac.rs]: Good call, I added the `CREATE` privilege and added a test in `privilege_checks.slt`

File: src/sql/src/plan/statement/ddl.rs
Comment (ParkMyCar) [reply src/sql/src/plan/statement/ddl.rs]: Added a comment, let me know if that suffices

File: src/adapter/src/coord/sequencer/inner.rs
Comment (ParkMyCar) [reply src/adapter/src/coord/sequencer/inner.rs]: Updated!

File: src/adapter/src/coord/command_handler.rs
Comment (ParkMyCar) [reply src/adapter/src/coord/command_handler.rs]: I expanded on this comment, the idea is to return a different type of error that an object was found, but it's the wrong type

File: src/environmentd/src/http/webhook.rs
Comment (ParkMyCar) [reply src/environmentd/src/http/webhook.rs:102]: I checked and it looks like the `http` crate already filters/modifies headers so they can all be represented as a `String`. For duplicate headers we retain the last one, I added a Rust test to assert this.

File: src/environmentd/src/http/webhook.rs
Comment (ParkMyCar) [reply src/environmentd/src/http/webhook.rs:169]: Yeah I struggled with the right error code here. The reason I didn't use 405 was this part:

> The server **must** generate an Allow header field in a 405 status code response. The field must contain a list of methods that the target resource currently supports.

I guess actually we could generate an empty Allow header?

415 could also work, but doesn't seem exactly right either because of this statement:

> The format problem might be due to the request's indicated [Content-Type](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type) or [Content-Encoding](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Encoding), or as a result of inspecting the data directly.

File: src/adapter/src/coord/sequencer/inner.rs
Comment (ParkMyCar) [reply src/adapter/src/coord/sequencer/inner.rs:199]: Added!

File: src/adapter/src/catalog.rs
Comment (ParkMyCar) [reply src/adapter/src/catalog.rs]: It doesn't appear to actually be called by anyone, and deleting the method returns no errors when I `cargo c --tests` the entire workspace 🤔

File: src/sql-parser/src/parser.rs
Comment (sploiselle) [reply src/sql-parser/src/parser.rs]: Ah, that makes a lot of sense. Thanks for providing the context.

File: src/sql/src/plan/statement/ddl.rs
Comment (sploiselle) [reply src/sql/src/plan/statement/ddl.rs:410]: This is good--ty

File: src/environmentd/src/http/webhook.rs
Comment (jkosh44) [reply src/environmentd/src/http/webhook.rs:169]: > I guess actually we could generate an empty Allow header?

Probably not worth it.

What about just [400 - Bad Request](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400) then? Otherwise, I think 403 is probably good enough.

**Review by jkosh44 — APPROVED** [review]
LGTM, though I'm not too familiar with the storage controller so I would also wait for a storage approval on that.

File: src/environmentd/src/http/webhook.rs
Comment (mjibson) [inline src/environmentd/src/http/webhook.rs:98]: Can this whole block be moved into the if so we don't allocate the map if we don't need to?

File: src/environmentd/src/http/webhook.rs
Comment (mjibson) [inline src/environmentd/src/http/webhook.rs:126]: This isn't a useful string for a user. It should change or move to `InternalStorageError`.

File: src/environmentd/tests/server.rs
Comment (mjibson) [inline src/environmentd/tests/server.rs:1700]: Use mz_ore::retry here instead.

**Review by mjibson — APPROVED** [review]
Looks great. I didn't review the storage-client stuff.

File: src/storage-client/src/controller.rs
Comment (sploiselle) [inline src/storage-client/src/controller.rs:1652]: We need to add the ability to remove these collections from the collection manager.

Complexities here include that right now, dropping a source only moves its `since` to `[]` and the function wants to be synchronous (i.e. non-blocking). Work that we want to handle asynchronously is done in the `process` fn in `src/storage-client/src/controller.rs`.

This source never makes it all the way to a Timely worker, so we'll need to manage more of the life cycle within the storage controller itself.

I'll be out when you're back, but Gus or Petros could help you out with this if a solution isn't obvious.

File: src/storage-client/src/controller/collection_mgmt.rs
Comment (sploiselle) [inline src/storage-client/src/controller/collection_mgmt.rs:156]: Just OOC, did you observe this message during your testing?

File: src/storage-client/src/controller/collection_mgmt.rs
Comment (sploiselle) [inline src/storage-client/src/controller/collection_mgmt.rs:119]: I think this should likely be its own error message; the case is not necessarily that the identifier is missing (though that's possible), but it might also have never been registered as a managed collection. Can't realistically know which so the error should reflect that in some way.

File: test/sqllogictest/webhook.slt
Comment (sploiselle) [inline test/sqllogictest/webhook.slt:102]: Ah, these tests typically only go in `src/sql-parser/tests/testdata/ddl` or `src/sql-parser/tests/testdata/create`. Looks like you already have some there, so let's just consolidate them to one spot.

File: src/adapter/src/coord/sequencer/inner.rs
Comment (sploiselle) [reply src/adapter/src/coord/sequencer/inner.rs:252]: SGTM

File: src/adapter/src/coord/command_handler.rs
Comment (sploiselle) [inline src/adapter/src/coord/command_handler.rs:827]: I think a generic 400 error is preferable; including any more detail than "you can't do that here" reveals more information about the catalog than is necessary.

**Review by sploiselle — COMMENTED** [review]
The storage stuff looks sane modulo the fact that we cannot remove collections. I'm not totally sure what the rollout plan for this is, but that feels important to tackle before letting anyone use this (e.g. these persist shards will live forever until we add the feature).

File: src/sql-parser/src/parser.rs
Comment (ParkMyCar) [reply src/sql-parser/src/parser.rs]: Done!

File: src/adapter/src/client.rs
Comment (ParkMyCar) [reply src/adapter/src/client.rs]: Updated!

File: src/adapter/src/client.rs
Comment (ParkMyCar) [reply src/adapter/src/client.rs]: We're only mapping to an internal error if receiving on the channel fails, other errors like a typo in the webhook source, are passed along. It's hard to see this though because it's done with the double `?`. I reformatted this a bit, hopefully that helps

File: src/adapter/src/coord/command_handler.rs
Comment (ParkMyCar) [reply src/adapter/src/coord/command_handler.rs:850]: I don't think we can because we don't create a Session when appending to a webhook, unless we want to use the system session?

File: src/sql/src/plan/statement/ddl.rs
Comment (ParkMyCar) [reply src/sql/src/plan/statement/ddl.rs:424]: Yes, when resolving a webhook source we soft_assert there are at most 2 columns

File: src/environmentd/src/http/webhook.rs
Comment (ParkMyCar) [reply src/environmentd/src/http/webhook.rs]: It does normalize capitalization for header names, [Rust Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=be0fae304295a8be046feb23ef188026)

File: src/environmentd/src/http/webhook.rs
Comment (ParkMyCar) [reply src/environmentd/src/http/webhook.rs]: Sounds good, removed!

File: src/environmentd/tests/server.rs
Comment (ParkMyCar) [reply src/environmentd/tests/server.rs]: Great idea, don't know why I didn't do this earlier

File: src/storage-client/src/controller.rs
Comment (ParkMyCar) [reply src/storage-client/src/controller.rs:1544]: Right! I'll do this work in a follow up PR since this one is already big enough, but I'll link it back here once it's up

File: src/storage-client/src/controller/collection_mgmt.rs
Comment (ParkMyCar) [reply src/storage-client/src/controller/collection_mgmt.rs:161]: I don't remember exactly, but I don't believe I did

File: test/sqllogictest/webhook.slt
Comment (ParkMyCar) [reply test/sqllogictest/webhook.slt]: I removed these to consolidate in `sql-parser`

File: src/adapter/src/coord/command_handler.rs
Comment (mjibson) [reply src/adapter/src/coord/command_handler.rs:850]: Right. That fn uses `resolve_entry` from a `CatalogState` underneath, which you do have access to. Can you use that?

Comment (ParkMyCar) [reply src/adapter/src/coord/command_handler.rs:850]: Ahhh I see, `resolve_entry(...)` is doing all of the manual logic I was doing myself. Updated!

**Pull request merged**
